torchtune/models/clip/_tokenizer.py

```diff
@@ -0,0 +1,206 @@
+        self.byte_decoder = {v: k for k, v in self.byte_encoder.items()}
+
+        merges = []
+        with open(path, encoding="utf-8") as f:
+            for i, line in enumerate(f):
+                line = line.strip()
+                if (i == 0 and line.startswith("#version:")) or not line:
+                    continue
+                merges.append(tuple(line.split()))
+
+        vocab = list(self.byte_encoder.values())
+        vocab.extend([v + "</w>" for v in vocab])
+        vocab.extend(["".join(merge) for merge in merges])
+        vocab.extend(["<|startoftext|>", "<|endoftext|>"])
+
+        self.encoder = {word: i for i, word in enumerate(vocab)}
+        self.decoder = {v: k for k, v in self.encoder.items()}
+        self.bpe_ranks = {merge: i for i, merge in enumerate(merges)}
+
+        self.pat = re.compile(
+            r"""<\|startoftext\|>|<\|endoftext\|>|'s|'t|'re|'ve|'m|'ll|'d|[\p{L}]+|[\p{N}]|[^\s\p{L}\p{N}]+""",
+            re.IGNORECASE,
+        )
+
+        self.sot_token = self.encoder["<|startoftext|>"]
+        self.eot_token = self.encoder["<|endoftext|>"]
+        self.pad_token = self.eot_token
+
+        self.cache = {
+            "<|startoftext|>": "<|startoftext|>",
+            "<|endoftext|>": "<|endoftext|>",
+        }
+
+    def encode(self, text: str) -> List[int]:
+        """
+        Given a string, return the encoded list of token ids.
+
+        Args:
+            text (str): The text to encode.
+
+        Returns:
+            List[int]: The encoded list of token ids.
+        """
+        text = text.lower()
+
+        tokens = [self.sot_token]
+        for token in re.findall(self.pat, text):
+            token = "".join(self.byte_encoder[b] for b in token.encode("utf-8"))
+            tokens.extend(
+                self.encoder[bpe_token] for bpe_token in self._bpe(token).split(" ")
+            )
+            if len(tokens) >= self.max_seq_len:
+                break
+        tokens.append(self.eot_token)
+
+        if len(tokens) > self.max_seq_len:
+            assert self.truncate, (
+                "Tokenized text is larger than the maximum sequence length but "
+                "truncate is set to False."
+            )
+            tokens = tokens[: self.max_seq_len]
+            tokens[-1] = self.eot_token
+
+        return tokens
+
+    def decode(self, tokens: List[int]) -> str:
+        """
+        Given a list of token ids, return the decoded text, optionally including special tokens.
+
+        Args:
+            tokens (List[int]): The list of token ids to decode.
+
+        Returns:
+            str: The decoded text.
+        """
+        text = "".join([self.decoder[token] for token in tokens])
+        return (
+            bytearray([self.byte_decoder[c] for c in text])
+            .decode("utf-8", errors="replace")
+            .replace("</w>", " ")
+        )
+
+    def __call__(self, texts: List[str]) -> torch.Tensor:
```

Comment:
In the case of diffusion, the text will be tokenized by both the CLIP tokenizer and the T5 tokenizer, so it probably makes sense to have the model handle this rather than the dataset. We won't be using the `SFTDataset` or `Message`s, we'll probably make new abstractions for image-text pairs.

I vote that we leave this as is for now, but revisit it once we decide on an overall design for our diffusion code. We might end up altering `__call__` or removing it entirely, but it's currently not clear

Reply:
The __call__ method is standardized and should look like this `def __call__(self, sample: Mapping[str, Any], inference: bool = False) -> Mapping[str, Any]:`. The inference flag doesn't have to do anything but should be there. As Rafi said, all tokenization happens in the dataset so it's not on the main execution path. For diffusion the data pipeline looks like
`Raw Dataset: (text, image) -> TuneDataset -> ModelTransform: {tokens, tensor} -> collation`